File: sando_py/nodes/auto_goal_pub.py
"""Auto terminal-goal publisher for the per-class demo.

sando_node's replan timer starts CANCELLED and only arms on the first term_goal,
so without a goal the drone never plans. This node publishes one PoseStamped on
`term_goal` a few seconds after launch (latched + repeated to beat startup
races), driving start -> goal with no RViz click.
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSDurabilityPolicy, QoSHistoryPolicy, QoSProfile, QoSReliabilityPolicy
from geometry_msgs.msg import PoseStamped


class AutoGoalPub(Node):
    def __init__(self):
        super().__init__("auto_goal_pub")
        self.declare_parameter("frame_id", "map")
        self.declare_parameter("goal_x", 10.0)        # end A
        self.declare_parameter("goal_y", 0.0)
        self.declare_parameter("goal_z", 2.0)
        self.declare_parameter("goal_x2", 0.0)        # end B (start) -> back-and-forth
        self.declare_parameter("goal_y2", 0.0)
        self.declare_parameter("goal_z2", 2.0)
        self.declare_parameter("delay_s", 3.0)
        self.declare_parameter("switch_period_s", 7.0)   # flip the goal every N s
        self.frame = self.get_parameter("frame_id").value
        self.delay = float(self.get_parameter("delay_s").value)
        self.switch_period = float(self.get_parameter("switch_period_s").value)
        self.targets = [
            (float(self.get_parameter("goal_x").value),
             float(self.get_parameter("goal_y").value),
             float(self.get_parameter("goal_z").value)),
            (float(self.get_parameter("goal_x2").value),
             float(self.get_parameter("goal_y2").value),
             float(self.get_parameter("goal_z2").value)),
        ]
        self.idx = 0

        qos = QoSProfile(history=QoSHistoryPolicy.KEEP_LAST, depth=10,
                         reliability=QoSReliabilityPolicy.RELIABLE,
                         durability=QoSDurabilityPolicy.TRANSIENT_LOCAL)
        self.pub = self.create_publisher(PoseStamped, "term_goal", qos)
        self._delay_timer = self.create_timer(self.delay, self._start)

    def _msg(self, g) -> PoseStamped:
        m = PoseStamped()
        m.header.stamp = self.get_clock().now().to_msg()
        m.header.frame_id = self.frame
        m.pose.position.x, m.pose.position.y, m.pose.position.z = g
        m.pose.orientation.w = 1.0
        return m

    def _start(self):
        self._delay_timer.cancel()
        self._send()                                       # first goal (end A)
        self.create_timer(self.switch_period, self._switch)   # then flip ends

    def _send(self):
        g = self.targets[self.idx]
        # republish a few times to beat any subscriber-up race
        for _ in range(3):
            self.pub.publish(self._msg(g))
        self.get_logger().info(f"term_goal -> {g}")

    def _switch(self):
        self.idx = (self.idx + 1) % 2                      # back-and-forth
        self._send()


def main(args=None):
    rclpy.init(args=args)
    node = AutoGoalPub()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: cpp/ros2/src/sando_cpp/src/auto_goal_pub.cpp
// auto_goal_pub.cpp — C++ ROS2 port of sando_py/nodes/auto_goal_pub.py.
// After `delay_s`, publishes a PoseStamped on `term_goal` (latched, 3x to beat races),
// then flips between end A / end B every `switch_period_s`.
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <array>
#include <chrono>
using namespace std::chrono_literals;

class AutoGoalPub : public rclcpp::Node {
 public:
  AutoGoalPub() : rclcpp::Node("auto_goal_pub") {
    frame_ = declare_parameter<std::string>("frame_id", "map");
    targets_[0] = {declare_parameter<double>("goal_x", 10.0), declare_parameter<double>("goal_y", 0.0),
                   declare_parameter<double>("goal_z", 2.0)};
    targets_[1] = {declare_parameter<double>("goal_x2", 0.0), declare_parameter<double>("goal_y2", 0.0),
                   declare_parameter<double>("goal_z2", 2.0)};
    delay_ = declare_parameter<double>("delay_s", 3.0);
    switch_period_ = declare_parameter<double>("switch_period_s", 7.0);

    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.reliable().transient_local();  // latched
    pub_ = create_publisher<geometry_msgs::msg::PoseStamped>("term_goal", qos);
    delay_timer_ = create_wall_timer(std::chrono::duration<double>(delay_), std::bind(&AutoGoalPub::start, this));
  }

 private:
  geometry_msgs::msg::PoseStamped msg(const std::array<double, 3>& g) {
    geometry_msgs::msg::PoseStamped m;
    m.header.stamp = now(); m.header.frame_id = frame_;
    m.pose.position.x = g[0]; m.pose.position.y = g[1]; m.pose.position.z = g[2];
    m.pose.orientation.w = 1.0;
    return m;
  }
  void send() {
    auto g = targets_[idx_];
    for (int i = 0; i < 3; ++i) pub_->publish(msg(g));
    RCLCPP_INFO(get_logger(), "term_goal -> [%.1f, %.1f, %.1f]", g[0], g[1], g[2]);
  }
  void start() {
    delay_timer_->cancel();
    send();
    switch_timer_ = create_wall_timer(std::chrono::duration<double>(switch_period_),
                                      std::bind(&AutoGoalPub::flip, this));
  }
  void flip() { idx_ = (idx_ + 1) % 2; send(); }

  std::string frame_;
  std::array<std::array<double, 3>, 2> targets_;
  double delay_, switch_period_;
  int idx_ = 0;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr delay_timer_, switch_timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AutoGoalPub>());
  rclcpp::shutdown();
  return 0;
}
